be/src/util/quantile_state.cpp

```diff
@@ -0,0 +1,377 @@
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+#include <string.h>
+#include <cmath>
+#include "util/quantile_state.h"
+#include "util/coding.h"
+#include "common/logging.h"
+
+
+namespace doris{
+
+template<typename T>
+QuantileState<T>::QuantileState():_type(EMPTY),compression(2048){}
+
+template<typename T>
+QuantileState<T>::QuantileState(float compression):_type(EMPTY),compression(compression){}
+
+
+template<typename T>
+QuantileState<T>::QuantileState(const Slice& slice) {
+    if (!deserialize(slice)) {
+        _type = EMPTY;
+    }
+}
+
+template<typename T>
+QuantileState<T>::~QuantileState() {
+    clear();
+}
+
+template<typename T>
+size_t QuantileState<T>::get_serialized_size() {
+    size_t size = 1 + sizeof(float); // type(QuantileStateType) + compression(float)
+    switch(_type) {
+    case EMPTY:
+        break;
+    case SINGLE:
+        size += sizeof(T);
+        break;
+    case EXPLICIT:
+        size += sizeof(uint16_t) + sizeof(T) * _explicit_data.size();
+        break;
+    case TDIGEST:
+        size += tdigest_ptr->serialized_size();
+        break;
+    }
+    return size;
+}
+
+template<typename T>
+void QuantileState<T>::set_compression(float compression) {
+    this->compression = compression;
+}
+
+template<typename T>
+bool QuantileState<T>::is_valid(const Slice& slice) {
+
+    if (slice.size < 1) {
+        return false;
+    }
+    const uint8_t* ptr = (uint8_t*)slice.data;
+    const uint8_t* end = (uint8_t*)slice.data + slice.size;
+    float compress_value = *reinterpret_cast<const float*>(ptr);
+    if (compress_value < 2048 || compress_value > 10000) {
+        return false;
+    }
+    ptr += sizeof(float);
+    
+
+    auto type = (QuantileStateType)*ptr++;
+    switch (type) {
+    case EMPTY:
+        break;
+    case SINGLE:{
+        if ((ptr + sizeof(T)) > end) {
+            return false;
+        }
+        ptr += sizeof(T);
+        // _single_data = *reinterpret_cast<T*>(ptr)++;
+        break;
+    }
+    case EXPLICIT: {
+        if ((ptr + sizeof(uint16_t)) > end) {
+            return false;
+        }
+        uint16_t num_explicits = decode_fixed16_le(ptr);
+        ptr += sizeof(uint16_t);
+        ptr += num_explicits * sizeof(T);
+        break;
+    }
+    case TDIGEST: {
+        if ((ptr + sizeof(uint32_t)) > end) {
+            return false;
+        }
+        uint32_t tdigest_serialized_length = decode_fixed32_le(ptr);
+        ptr += tdigest_serialized_length;
```

Comment:
Why not check content of tdigest? Such as ```isDirty()```.

Reply:
Here just check whether the total length is correct.
If check the content , we should call `deserialize()` first. It will take lots of time.